design/automatic-network-configuration.md

```diff
@@ -0,0 +1,738 @@
+  within the same namespace.
+
+### Permissions
+
+The permissions for a role include the permissions given to the controllers
+deployed by that role.
+
+#### Scenario: One cluster operator
+
+Description: Only one cluster operator has full management access to all resources
+
+Multi-tenancy models:
+
+- None
+
+Roles:
+
+- Operator: Full control
+  Scope: Cluster
+
+Operations Allowed:
+||Operation / resource|Goal|
+|:-|:-|:-|
+|Operator|All/All|Configure everything|
+
+Operations Not Allowed:
+||Operation|Goal|
+|:-|:-|:-|
+|Operator|None||
+
+#### Scenario: Many cluster operators
+
+Description: Each operator has full management access to all resources in his namespace.
+
+Multi-tenancy models:
+
+- Full
+
+Roles:
+
+- Operator: Full control
+  Scope: Cluster
+
+Operations Allowed:
+||Operation / resource|Goal|
+|:-|:-|:-|
+|Operator|All/All|Configure everything|
+
+Operations Not Allowed:
+||Operation|Goal|
+|:-|:-|:-|
+|Operator|None||
+
+#### Scenario: Infra team and consumers
+
+Description: There is an owner of the infra and some consumers. The infra
+team is the owner of the management cluster and manages the physical resources.
+A consumer is the manager of the target cluster, responsible for its lifecycle.
+The consumer gets resources from the infra team (tenant). No consumer action
+should impact another consumer. Metal3 setup on a shared management cluster,
+with namespaces for each consumer.
+
+The switch is shared between multiple consumers (dedicated ports, vlans etc.)
+and managed by the infra team.
+
+Multi-tenancy models:
+
+- Limited
+
+Roles:
+
+- User: Configuration of existing ports
+  Scope: Namespaced
+- Operator: Definition of the infra
+  Scope: Namespaced
+
+Operations Allowed:
+||Operation/resource|Goal|
+|:-|:-|:-|
+|User|Create-delete-Update/SwitchPortConfiguration|Create any configuration|
+|User|Modify the SwitchPort|Link a configuration|
+|Operator|All/All||
+
+Operations Not Allowed:
+||Operation/resource|Goal|
+|:-|:-|:-|
+|User|Create-Delete/SwitchPort|Modify the infra objects|
+|User|All/Switch|Modify the infra objects|
+|Operator|None||
+
+### Implementation Details/Notes/Constraints
+
+#### Re-applying the configuration (no configuration change in K8S)
+
+Pros:
+
+- Declarative approach, keep in sync the K8S object and the switch
+  configuration.
+- React on errors / misconfigurations etc.
+- Trap configuration changes between reconciliation.
+
+Cons:
+
+- Configurability might be missing some coverage, operator manually changes it.
+  Some exceptions might be reasonable (e.g. credentials change)
+- Might be heavy process (can be fixed by reducing the frequency,
+  jittering etc.)
+- Other actors might increase complexity.
+- No full ownership guaranteed of the hardware
+- Could break during an upgrade
+
+Possibility to add a boolean in the CR to control whether we reconcile and
+re-apply the configuration if there is a delta, or not.
```

Comment:
IMHO it's a mistake to make it configurable. I'd suggest you always re-apply, but start with a very long polling interval if you're worried about resource usage.
Don't let people get used to the idea that they can manually change the config behind the controller's back and not have it reverted.

Reply:
got it, thanks!